### PR TITLE
NGP conversion of TpetraLinearSystem::applyDirichletBCs

### PR DESCRIPTION
**Pull request by alanw0**
**Pull-request type:**
- [ ] Bug fix 
- [ ] Documentation update
- [x ] Feature enhancement

## Description of the pull-request

NGP conversion of TpetraLinearSystem::applyDirichletBCs with unit-test in unit_tests/edge_kernels/UnitTestScalarAdvDiffEdge.C

## Checklist

**NOTE** The checklist below is a suggestion and not mandatory. You don't have
to _check all boxes_ to submit a pull request. However, please add a brief
explanation in your pull request summary explaining the omission for the benefit
of reviewers and developers.

*All pull requests*
- [ ] Builds successfully (*must test on at least one system/compiler combination*)
  - Operating systems 
    - [x ] Linux
    - [ ] MacOS
  - Compilers 
    - [ ] GCC
    - [ ] LLVM/Clang
    - [ ] Intel compilers
    - [x ] NVIDIA CUDA
- [ ] Compiles without warnings
- [ ] Passes all unit tests
- [ ] Passes all regression tests
- [ ] Documentation builds without errors

*For new code updates*
- [ ] Documentation updates - additions to user, theory, and verification manuals
- [ ] New unit tests providing coverage for new code, bug fixes etc.
- [ ] New regression tests exercising the new code

### CODE REVIEW

File: src/TpetraLinearSystem.C [hunk 1]
```diff
@@ -1474,64 +1492,50 @@ void TpetraLinearSystem::applyDirichletBCs(stk::mesh::FieldBase * solutionField,
...
-    ThrowRequire(fieldSize == numDof_);
+  ngp::Mesh ngpMesh = realm_.ngp_mesh();
+  NGPDoubleFieldType ngpSolutionField = realm_.ngp_field_manager().get_field<double>(solutionField->mesh_meta_data_ordinal());
+  NGPDoubleFieldType ngpBCValuesField = realm_.ngp_field_manager().get_field<double>(bcValuesField->mesh_meta_data_ordinal());
 
-    const stk::mesh::Bucket::size_type length   = b.size();
-    const double * solution = (double*)stk::mesh::field_data(*solutionField, *b.begin());
-    const double * bcValues = (double*)stk::mesh::field_data(*bcValuesField, *b.begin());
+  ngpSolutionField.sync_to_device();
+  ngpBCValuesField.sync_to_device();
```
Comment (sayerhs) [inline src/TpetraLinearSystem.C]: Is this just to be safe?

File: src/TpetraLinearSystem.C [hunk 2]
```diff
@@ -1474,64 +1492,50 @@ void TpetraLinearSystem::applyDirichletBCs(stk::mesh::FieldBase * solutionField,
...
-        const LinSys::LocalMatrix& local_matrix = useOwned ? ownedLocalMatrix_ : sharedNotOwnedLocalMatrix_;
-
-        if(localId > maxSharedNotOwnedRowId_) {
-          std::cerr << "localId > maxSharedNotOwnedRowId_:: localId= " << localId << " maxSharedNotOwnedRowId_= " << maxSharedNotOwnedRowId_ << " useOwned = " << (localId < maxOwnedRowId_ ) << std::endl;
-          throw std::runtime_error("logic error: localId > maxSharedNotOwnedRowId_");
-        }
+        const LocalOrdinal actualLocalId = useOwned ? localId : localId - maxOwnedRowId;
 
-        // Adjust the LHS
+        NGP_ThrowRequire(localId <= maxSharedNotOwnedRowId);
```
Comment (sayerhs) [inline src/TpetraLinearSystem.C]: Should this be `NGP_ThrowAssert`? Wanted to understand the performance implications.

File: src/TpetraLinearSystem.C (hunk 1 above)
Comment (alanw0) [reply src/TpetraLinearSystem.C]: Yes. I just didn't want to count on the fields already being sync'd. If they are already sync'd, these sync_to_device calls should be very fast.

File: src/TpetraLinearSystem.C (hunk 2 above)
Comment (alanw0) [reply src/TpetraLinearSystem.C]: Yes NGP_ThrowAssert would probably be better here. I'll add a commit and do that.

**Pull request merged**